File: app/src/main/java/com/zeapo/pwdstore/autofill/AutofillRecyclerAdapter.java
package com.zeapo.pwdstore.autofill;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SortedList;
import androidx.recyclerview.widget.SortedListAdapterCallback;
import com.zeapo.pwdstore.R;

import java.util.ArrayList;
import java.util.List;

class AutofillRecyclerAdapter extends RecyclerView.Adapter<AutofillRecyclerAdapter.ViewHolder> {

    private SortedList<AppInfo> apps;
    private ArrayList<AppInfo> allApps; // for filtering, maintain a list of all
    private AutofillPreferenceActivity activity;
    private Drawable browserIcon = null;

    AutofillRecyclerAdapter(List<AppInfo> allApps, final PackageManager pm
            , AutofillPreferenceActivity activity) {
        SortedList.Callback<AppInfo> callback = new SortedListAdapterCallback<AppInfo>(this) {
            // don't take into account secondary text. This is good enough
            // for the limited add/remove usage for websites
            @Override
            public int compare(AppInfo o1, AppInfo o2) {
                return o1.appName.toLowerCase().compareTo(o2.appName.toLowerCase());
            }

            @Override
            public boolean areContentsTheSame(AppInfo oldItem, AppInfo newItem) {
                return oldItem.appName.equals(newItem.appName);
            }

            @Override
            public boolean areItemsTheSame(AppInfo item1, AppInfo item2) {
                return item1.appName.equals(item2.appName);
            }
        };
        this.apps = new SortedList<>(AppInfo.class, callback);
        this.apps.addAll(allApps);
        this.allApps = new ArrayList<>(allApps);
        this.activity = activity;
        try {
            browserIcon = activity.getPackageManager().getApplicationIcon("com.android.browser");
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
    }

    @Override
    public AutofillRecyclerAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.autofill_row_layout, parent, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(AutofillRecyclerAdapter.ViewHolder holder, int position) {
        AppInfo app = apps.get(position);
        holder.packageName = app.packageName;
        holder.appName = app.appName;
        holder.isWeb = app.isWeb;

        holder.icon.setImageDrawable(app.icon);
        holder.name.setText(app.appName);

        holder.secondary.setVisibility(View.VISIBLE);
        holder.view.setBackgroundResource(R.color.grey_white_1000);

        SharedPreferences prefs;
        if (!app.appName.equals(app.packageName)) {
            prefs = activity.getApplicationContext().getSharedPreferences("autofill", Context.MODE_PRIVATE);
        } else {
            prefs = activity.getApplicationContext().getSharedPreferences("autofill_web", Context.MODE_PRIVATE);
        }
        String preference = prefs.getString(holder.packageName, "");
        switch (preference) {
            case "":
                holder.secondary.setVisibility(View.GONE);
                holder.view.setBackgroundResource(0);
                break;
            case "/first":
                holder.secondary.setText(R.string.autofill_apps_first);
                break;
            case "/never":
                holder.secondary.setText(R.string.autofill_apps_never);
                break;
            default:
                holder.secondary.setText(R.string.autofill_apps_match);
                holder.secondary.append(" " + preference.split("\n")[0]);
                if ((preference.trim().split("\n").length - 1) > 0) {
                    holder.secondary.append(" and "
                            + (preference.trim().split("\n").length - 1) + " more");
                }
                break;
        }
    }

    @Override
    public int getItemCount() {
        return apps.size();
    }

    int getPosition(String appName) {
        return apps.indexOf(new AppInfo(null, appName, false, null));
    }

    // for websites, URL = packageName == appName
    void addWebsite(String packageName) {
        apps.add(new AppInfo(packageName, packageName, true, browserIcon));
        allApps.add(new AppInfo(packageName, packageName, true, browserIcon));
    }

    void removeWebsite(String packageName) {
        apps.remove(new AppInfo(null, packageName, false, null));
        allApps.remove(new AppInfo(null, packageName, false, null)); // compare with equals
    }

    void updateWebsite(String oldPackageName, String packageName) {
        apps.updateItemAt(getPosition(oldPackageName), new AppInfo(packageName, packageName, true, browserIcon));
        allApps.remove(new AppInfo(null, oldPackageName, false, null)); // compare with equals
        allApps.add(new AppInfo(null, packageName, false, null));
    }

    void filter(String s) {
        if (s.isEmpty()) {
            apps.addAll(allApps);
            return;
        }
        apps.beginBatchedUpdates();
        for (AppInfo app : allApps) {
            if (app.appName.toLowerCase().contains(s.toLowerCase())) {
                apps.add(app);
            } else {
                apps.remove(app);
            }
        }
        apps.endBatchedUpdates();
    }

    static class AppInfo {
        public Drawable icon;
        String packageName;
        String appName;
        boolean isWeb;

        AppInfo(String packageName, String appName, boolean isWeb, Drawable icon) {
            this.packageName = packageName;
            this.appName = appName;
            this.isWeb = isWeb;
            this.icon = icon;
        }

        @Override
        public boolean equals(Object o) {
            return o != null && o instanceof AppInfo && this.appName.equals(((AppInfo) o).appName);
        }
    }

    class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        public View view;
        public TextView name;
        public ImageView icon;
        TextView secondary;
        String packageName;
        String appName;
        Boolean isWeb;

        ViewHolder(View view) {
            super(view);
            this.view = view;
            name = (TextView) view.findViewById(R.id.app_name);
            secondary = (TextView) view.findViewById(R.id.secondary_text);
            icon = (ImageView) view.findViewById(R.id.app_icon);
            view.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            activity.showDialog(packageName, appName, isWeb);
        }

    }
}

File: app/src/main/java/com/zeapo/pwdstore/autofill/AutofillRecyclerAdapter.kt
package com.zeapo.pwdstore.autofill

import android.content.Context
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.SortedList
import androidx.recyclerview.widget.SortedListAdapterCallback
import com.zeapo.pwdstore.R
import java.util.ArrayList
import java.util.Locale

internal class AutofillRecyclerAdapter(
        allApps: List<AppInfo>,
        private val activity: AutofillPreferenceActivity
) : RecyclerView.Adapter<AutofillRecyclerAdapter.ViewHolder>() {

    private val apps: SortedList<AppInfo>
    private val allApps: ArrayList<AppInfo> // for filtering, maintain a list of all
    private var browserIcon: Drawable? = null

    init {
        val callback = object : SortedListAdapterCallback<AppInfo>(this) {
            // don't take into account secondary text. This is good enough
            // for the limited add/remove usage for websites
            override fun compare(o1: AppInfo, o2: AppInfo): Int {
                return o1.appName.toLowerCase(Locale.ROOT).compareTo(o2.appName.toLowerCase(Locale.ROOT))
            }

            override fun areContentsTheSame(oldItem: AppInfo, newItem: AppInfo): Boolean {
                return oldItem.appName == newItem.appName
            }

            override fun areItemsTheSame(item1: AppInfo, item2: AppInfo): Boolean {
                return item1.appName == item2.appName
            }
        }
        apps = SortedList(AppInfo::class.java, callback)
        apps.addAll(allApps)
        this.allApps = ArrayList(allApps)
        try {
            browserIcon = activity.packageManager.getApplicationIcon("com.android.browser")
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.autofill_row_layout, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val app = apps.get(position)
        holder.packageName = app.packageName
        holder.appName = app.appName
        holder.isWeb = app.isWeb

        holder.icon.setImageDrawable(app.icon)
        holder.name.text = app.appName

        holder.secondary.visibility = View.VISIBLE
        holder.view.setBackgroundResource(R.color.grey_white_1000)

        val prefs: SharedPreferences
        prefs = if (app.appName != app.packageName) {
            activity.applicationContext.getSharedPreferences("autofill", Context.MODE_PRIVATE)
        } else {
            activity.applicationContext.getSharedPreferences("autofill_web", Context.MODE_PRIVATE)
        }
        when (val preference = prefs.getString(holder.packageName, "")) {
            "" -> {
                holder.secondary.visibility = View.GONE
                holder.view.setBackgroundResource(0)
            }
            "/first" -> holder.secondary.setText(R.string.autofill_apps_first)
            "/never" -> holder.secondary.setText(R.string.autofill_apps_never)
            else -> {
                holder.secondary.setText(R.string.autofill_apps_match)
                holder.secondary.append(" " + preference!!.splitLines()[0])
                if (preference.trim { it <= ' ' }.splitLines().size - 1 > 0) {
                    holder.secondary.append(" and "
                            + (preference.trim { it <= ' ' }.splitLines().size - 1) + " more")
                }
            }
        }
    }

    override fun getItemCount(): Int {
        return apps.size()
    }

    fun getPosition(appName: String): Int {
        return apps.indexOf(AppInfo(null, appName, false, null))
    }

    private fun String.splitLines(): Array<String> {
        return split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
    }

    // for websites, URL = packageName == appName
    fun addWebsite(packageName: String) {
        apps.add(AppInfo(packageName, packageName, true, browserIcon))
        allApps.add(AppInfo(packageName, packageName, true, browserIcon))
    }

    fun removeWebsite(packageName: String) {
        apps.remove(AppInfo(null, packageName, false, null))
        allApps.remove(AppInfo(null, packageName, false, null)) // compare with equals
    }

    fun updateWebsite(oldPackageName: String, packageName: String) {
        apps.updateItemAt(getPosition(oldPackageName), AppInfo(packageName, packageName, true, browserIcon))
        allApps.remove(AppInfo(null, oldPackageName, false, null)) // compare with equals
        allApps.add(AppInfo(null, packageName, false, null))
    }

    fun filter(s: String) {
        if (s.isEmpty()) {
            apps.addAll(allApps)
            return
        }
        apps.beginBatchedUpdates()
        for (app in allApps) {
            if (app.appName.toLowerCase(Locale.ROOT).contains(s.toLowerCase(Locale.ROOT))) {
                apps.add(app)
            } else {
                apps.remove(app)
            }
        }
        apps.endBatchedUpdates()
    }

    internal class AppInfo(var packageName: String?, var appName: String, var isWeb: Boolean, var icon: Drawable?) {

        override fun equals(other: Any?): Boolean {
            return other is AppInfo && this.appName == other.appName
        }

        override fun hashCode(): Int {
            var result = packageName?.hashCode() ?: 0
            result = 31 * result + appName.hashCode()
            result = 31 * result + isWeb.hashCode()
            result = 31 * result + (icon?.hashCode() ?: 0)
            return result
        }
    }

    internal inner class ViewHolder(var view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {
        var name: TextView = view.findViewById(R.id.app_name)
        var icon: ImageView = view.findViewById(R.id.app_icon)
        var secondary: TextView = view.findViewById(R.id.secondary_text)
        var packageName: String? = null
        var appName: String? = null
        var isWeb: Boolean = false

        init {
            view.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            activity.showDialog(packageName, appName, isWeb)
        }

    }
}
